### ISSUE TITLE
2.1.9 Regression, NoSuchMethodError com.github.jasync.sql.db.Configuration

### ISSUE THREAD

**Issue by KevinAtSesam** [jasync-sql/jasync-sql#359]
```
{
  "level": "ERROR",
  "title": "Uncaught exception",
  "origin": "Main",
  "exception": [
    "java.lang.NoSuchMethodError: 
      'com.github.jasync.sql.db.Configuration com.github.jasync.sql.db.Configuration.copy(java.lang.String, java.lang.String, int, java.lang.String, java.lang.String, com.github.jasync.sql.db.SSLConfiguration, java.nio.charset.Charset, int, io.netty.buffer.ByteBufAllocator, int, java.time.Duration, java.lang.String, java.util.List, io.netty.channel.EventLoopGroup, java.util.concurrent.Executor, java.lang.String, java.lang.String)'
  ]
}
```

**Comment by oshai** [comment]
Can you please provide some more details?

**Comment by KevinAtSesam** [comment]
Of course, let me know what you need.

Anyway, we use a series of libraries in conjunction with jasync-sql . For one, we use [scalikejdbc-async](https://github.com/scalikejdbc/scalikejdbc-async) **0.16.0** for writing our queries and such. When we initialize the database, we we a connection pool.
```
AsyncConnectionPool.add(
  postgresConfig.name,
  postgresConfig.url,
  postgresConfig.user,
  postgresConfig.pass,
  asyncConnectionPoolSettings
)

[...]

  lazy val asyncConnectionPoolSettings: AsyncConnectionPoolSettings = AsyncConnectionPoolSettings(
    // Must be lower then the RDS connection limit https://serverfault.com/a/982173
    maxPoolSize = 900,
    // Total amount of requests that can be held before being executed
    maxQueueSize = 10000,
    // Number of milliseconds for which the objects are going to be kept as idle
    maxIdleMillis = 1000L,
    connectionSettings = AsyncConnectionSettings(
      // Time allowed between our platform and setting up a connection to Postgres
      connectTimeout = Some(5.seconds),
      // Time for testing connection pools
      testTimeout = Some(5.seconds)
      // Uncertain if we should use this
      // queryTimeout = Some(20.seconds)
    )
  )
```

My guess is that one of the settings are not properly passed on. 

## Version kerfuffle
We are currently on jasync-postgresql 2.1.8 and scalikejdbc-async 0.16.0, and all works well, but the moment I upgrade the jasync library to 2.1.9, there are errors when starting the database connection.

**Comment by oshai** [comment]
I think this is a compatibility issue with https://github.com/jasync-sql/jasync-sql/pull/335

What will be helpful is a full stacktrace (whats calling that missing method).

**Comment by oshai** [comment]
I am pretty sure the issue is that using copy is a breaking change, and we can't override it. See: https://kotlinlang.org/docs/data-classes.html

The solution would be to provide a hand-tailored copy method and use that.

**Comment by oshai** [comment]
I created a fix here: https://github.com/jasync-sql/jasync-sql/pull/362, but since it's a breaking change it will probably also require a fix in scalike-jdbc.

In order to avoid that in the future, I am thinking to change configuration class to non data class. but since it's also a breaking change let's start with current PR.

**Comment by oshai** [comment]
Also opened a youtrack issue here: https://youtrack.jetbrains.com/issue/KT-56022/Support-JvmOverload-for-data-class-copy-function

**Comment by oshai** [comment]
I am also trying a different approach, to replace the data class with in class impl.

**Comment by oshai** [comment]
I release 2.1.15 with a fix and it should be available soon.  
I changed relevant copy methods (not using data class) so I hope it will solve the binary compatibility, although I am not 100% sure.  
Please try the new version and update.

**Comment by oshai** [comment]
I skipped 2.1.15 because of a flaky test. So the change will be in 2.1.16.

**Comment by KevinAtSesam** [comment]
I did some testing but to no avail;

> Uncaught exception when running com.sesamsolutions.platform.accountmgmt.api.VerificationApiSpec: java.lang.NoSuchMethodError: 'com.github.jasync.sql.db.Configuration com.github.jasync.sql.db.Configuration.copy(java.lang.String, java.lang.String, int, java.lang.String, java.lang.String, com.github.jasync.sql.db.SSLConfiguration, java.nio.charset.Charset, int, io.netty.buffer.ByteBufAllocator, int, java.time.Duration, java.lang.String, java.util.List, io.netty.channel.EventLoopGroup, java.util.concurrent.Executor, java.lang.String, java.lang.String)'

Tested this with scalikejdbc-async 0.16.0 and 0.17.0 so it's not resolved yet. Is there more I can do to help you troubleshoot?

**Comment by oshai** [comment]
Comparing signatures and looking at the bytecode the difference is that integers param are int instead of Integer. 
I will try something else.

**Comment by oshai** [comment]
Please try version 2.1.19 when available.

**Comment by KevinAtSesam** [comment]
With 2.1.19 I had this;

>  java.lang.NoSuchMethodError: 'com.github.jasync.sql.db.Configuration com.github.jasync.sql.db.Configuration.copy(java.lang.String, java.lang.String, int, java.lang.String, java.lang.String, com.github.jasync.sql.db.SSLConfiguration, java.nio.charset.Charset, int, io.netty.buffer.ByteBufAllocator, int, java.time.Duration, java.lang.String, java.util.List, io.netty.channel.EventLoopGroup, java.util.concurrent.Executor, java.lang.String, java.lang.String)'

I tried 2.1.21 and it gave me; 

> java.lang.NoSuchMethodError: 'com.github.jasync.sql.db.Configuration com.github.jasync.sql.db.Configuration.copy(java.lang.String, java.lang.String, int, java.lang.String, java.lang.String, com.github.jasync.sql.db.SSLConfiguration, java.nio.charset.Charset, int, io.netty.buffer.ByteBufAllocator, int, java.time.Duration, java.lang.String, java.util.List, io.netty.channel.EventLoopGroup, java.util.concurrent.Executor, java.lang.String, java.lang.String)'

**Comment by oshai** [comment]
Adding here scalike issue for ref: https://github.com/scalikejdbc/scalikejdbc-async/issues/381

**Comment by oshai** [comment]
> With 2.1.19 I had this;
> 
> > java.lang.NoSuchMethodError: 'com.github.jasync.sql.db.Configuration com.github.jasync.sql.db.Configuration.copy(java.lang.String, java.lang.String, int, java.lang.String, java.lang.String, com.github.jasync.sql.db.SSLConfiguration, java.nio.charset.Charset, int, io.netty.buffer.ByteBufAllocator, int, java.time.Duration, java.lang.String, java.util.List, io.netty.channel.EventLoopGroup, java.util.concurrent.Executor, java.lang.String, java.lang.String)'
> 
> I tried 2.1.21 and it gave me;
> 
> > java.lang.NoSuchMethodError: 'com.github.jasync.sql.db.Configuration com.github.jasync.sql.db.Configuration.copy(java.lang.String, java.lang.String, int, java.lang.String, java.lang.String, com.github.jasync.sql.db.SSLConfiguration, java.nio.charset.Charset, int, io.netty.buffer.ByteBufAllocator, int, java.time.Duration, java.lang.String, java.util.List, io.netty.channel.EventLoopGroup, java.util.concurrent.Executor, java.lang.String, java.lang.String)'

Does it happen on runtime or compilation?

**Comment by oshai** [comment]
Found one more issue with the signature, fixed in 2.1.23 so please try that version.

**Comment by KevinAtSesam** [comment]
> Does it happen on runtime or compilation?
All these issues happen on runtime. We have quite a large test-suite so it's easy to reproduce and check.

> Found one more issue with the signature, fixed in 2.1.23 so please try that version.
This does it! Seems that you've done it!

(Double checked with  scalikejdbc-async 0.17.0 and it worked too)

**Comment by oshai** [comment]
Great! closing this issue.
